JP-2152: Add an upper tweak threshold to tweakreg

jdavies-st opened the pull request:
Resolves #6154 / [JP-2152](https://jira.stsci.edu/browse/JP-2152)

**Description**

This PR adds an upper tweak threshold to `TweakregStep`.  I.e. any tweaks larger than 10 arcsec are probably due to poor data, poor catalogs, or both, and should be rejected.

This hard codes the threshold to be 10 arcsec.  But honestly, we could also use the existing `self.threshold` parameter in tweakreg of 1.0 arcsec as well.  Thoughts?  I don't want to add a separate parameter yet, as I don't think this is something users need to play with right now.  There are already too many knobs in this step to play with.

Checklist
- [ ] Tests
- [ ] Documentation
- [x] Change log
- [x] Milestone
- [x] Label(s)

hbushouse reviewed (approved):
Code looks good. In terms of how to implement the upper threshold, if you don't feel well about hard coding the 10 arcsec limit, but at the same time don't want to add yet another step param, what about doing something like 10 * `self.threshold`?

mcara commented on jwst/tweakreg/tweakreg_step.py:
```diff
@@ -300,6 +312,28 @@ def process(self, input):
+        wcs = imcat.meta['image_model'].meta.wcs
+        twcs = imcat.wcs
+
+        grid = grid_from_bounding_box(wcs.bounding_box, step=100)
+        # There's a bug in grid_from_bounding_box that makes the final grid
+        # item outside the bounding box.  So trim off the last one.  And trim
+        # the first one to avoid edge effects.
+        grid = grid[:,1:-1,1:-1]
```
I think this is an overkill. Using the corners of the bounding box should be enough: these are extreme points of the image and should reveal large shift, scale, and/or rotation.

jdavies-st replied:
Great feedback.  `wcs.footprint().T` gives just the bounding box corners in RA/Dec, so I'll just use those.

mcara reviewed (approved):
I approve this PR because it should do what it is supposed to do. However, for performance reasons, I highly recommend simplifying the grid to just 4 corners of the bounding box. I would be surprised if inner pixels can have larger separation than the largest separation of the 4 corners.

jdavies-st commented:
I used `10 * self.tolerance` as you suggested @hbushouse .  We can reduce the multiplicative factor in the future if needed.

I changed from using the a grid of points to just the corners, i.e. `wcs.footprint().T` as you suggested @mcara .  This is much better and clearer.

Finally I added a unit test to make sure the function is working as expected, and doing so made me change the calling signature to be a bit more test-friendly.

nden commented on jwst/tweakreg/tweakreg_step.py:
```diff
@@ -312,21 +313,12 @@ def process(self, input):
+        tolerance = 10.0 * self.tolerance * u.arcsec
 
+        ra, dec = wcs.footprint(axis_type="spatial").T
+        tra, tdec = twcs.footprint(axis_type="spatial").T
```
I think this would be the right approach but is it expected to work with spectral resampling? The Nirspec slits will produce NaNs as a footprint. Somewhere in the jwst code there might be a function to deal with this but I'm not sure and will have to look fo rit.

hbushouse replied:
tweakreg isn't used for spectra, only images

nden replied:
Duh!
